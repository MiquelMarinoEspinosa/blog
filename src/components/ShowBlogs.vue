<template>
  <div id="show-blogs">
    <h1>All Blog Articles</h1>
    <input type="text" v-model="search" placeholder="search blogs" />
    <div class="single-blog" v-for='(blog, index) in filteredBlogs' :key=index>
        <router-link v-bind:to="'/blog/' + blog.id"><h2 v-rainbow>{{ blog.title | to-uppercase }}</h2></router-link>
        <article>{{ blog.body | snippet }}</article>
    </div>
  </div>
</template>

<script>
import SearchMixin from '../mixins/SearchMixin'

export default {
    data() {
        return {
            blogs: [],
            search: ''
        }
    },
    methods: {

    },
    created() {
        this.$http.get('https://jsonplaceholder.typicode.com/posts').then((response) => {
            this.blogs = response.data.slice(0,10);
        });
    },
    computed: {
        
    },
    filters: {
        toUppercase(value) {
            return value.toUpperCase();
        }
    },
    directives: {
        'rainbow': {
            bind(el){
                el.style.color = "#" + Math.random().toString().slice(2,8);
            }
        }
    },
    mixins: [SearchMixin]
}
</script>

<style>
#show-blogs{
    max-width: 800px;
    margin: 0 auto;
}
.single-blog{
    padding: 20px;
    margin: 20px 0;
    box-sizing: border-box;
    background: #eee;
}
</style>