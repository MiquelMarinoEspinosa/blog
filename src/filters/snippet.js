export default (value) => {
  return value ? value.slice(0, 100) + "..." : value;
};
