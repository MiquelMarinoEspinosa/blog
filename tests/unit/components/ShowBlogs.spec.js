import { shallowMount, createLocalVue } from "@vue/test-utils";
import ShowBlogs from "@/components/ShowBlogs";
import Router from "vue-router";
import axios from "axios";
import snippet from "@/filters/snippet";

const localVue = createLocalVue();
localVue.prototype.$http = axios;
localVue.use(Router);
localVue.filter("snippet", snippet);

jest.mock("axios");

describe("Showblogs list of blogs", () => {
  let showBlogs;

  beforeEach(() => {
    jest.clearAllMocks();
  });

  afterEach(() => {
    showBlogs = null;
  });

  it("renders an empty list of blogs", async () => {
    const data = [];
    axios.get.mockResolvedValue({ data });
    showBlogs = shallowMount(ShowBlogs, { localVue });
    await localVue.nextTick();
    expect(axios.get).toHaveBeenCalledTimes(1);
    expect(axios.get).toHaveBeenCalledWith("http://localhost:8081/blog");
    expect(showBlogs.html()).toMatchSnapshot();
  });

  it("search by title into an empty blogs list returns an empty list", async () => {
    axios.get.mockResolvedValue({ data: [] });
    showBlogs = shallowMount(ShowBlogs, { localVue });
    await localVue.nextTick();
    const searchBox = showBlogs.get("input[type=text]");
    searchBox.setValue("Blog 1");
    await localVue.nextTick();
    expect(showBlogs.html()).toMatchSnapshot();
  });

  it("renders a list with some blogs entries", async () => {
    const data = [
      {
        title: "Blog 1",
        content: "Blog content 1",
        author: "Author 1",
        categories: ["vue", "react"],
      },
      {
        title: "Blog 2",
        content: "Blog content 2",
        author: "Author 2",
        categories: ["vue"],
      },
      {
        title: "Blog 3",
        content: "Blog content 3",
        author: "Author 2",
        categories: ["react"],
      },
    ];

    axios.get.mockResolvedValue({ data });
    showBlogs = shallowMount(ShowBlogs, { localVue });
    await localVue.nextTick();
    expect(axios.get).toHaveBeenCalledTimes(1);
    expect(axios.get).toHaveBeenCalledWith("http://localhost:8081/blog");
    expect(showBlogs.html()).toMatchSnapshot();
  });

  it("search by title into an empty blogs list returns list with 2 items", async () => {
    const data = [
      {
        title: "Another Blog 1",
        content: "Blog content 1",
        author: "Author 1",
        categories: ["vue", "react"],
      },
      {
        title: "Another Blog 2",
        content: "Blog content 2",
        author: "Author 2",
        categories: ["vue"],
      },
      {
        title: "Blog 3",
        content: "Blog content 3",
        author: "Author 2",
        categories: ["react"],
      },
    ];
    axios.get.mockResolvedValue({ data });
    showBlogs = shallowMount(ShowBlogs, { localVue });
    await localVue.nextTick();
    const searchBox = showBlogs.get("input[type=text]");
    searchBox.setValue("Another");
    await localVue.nextTick();
    expect(showBlogs.html()).toMatchSnapshot();
  });
});
