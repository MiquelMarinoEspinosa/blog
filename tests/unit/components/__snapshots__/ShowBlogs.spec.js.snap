// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Showblogs list of blogs renders an empty list of blogs 1`] = `
<div id="show-blogs">
  <h1>All Blog Articles</h1> <input type="text" placeholder="search blogs">
</div>
`;

exports[`Showblogs list of blogs search by title into an empty blogs list returns an empty list 1`] = `
<div id="show-blogs">
  <h1>All Blog Articles</h1> <input type="text" placeholder="search blogs">
</div>
`;

exports[`Showblogs list of blogs renders a list with some blogs entries 1`] = `
<div id="show-blogs">
  <h1>All Blog Articles</h1> <input type="text" placeholder="search blogs">
  <div class="single-blog">
    <router-link-stub to="/blog/undefined" tag="a" ariacurrentvalue="page" event="click">
      <h2>BLOG 1</h2>
    </router-link-stub>
    <article>Blog content 1...</article>
  </div>
  <div class="single-blog">
    <router-link-stub to="/blog/undefined" tag="a" ariacurrentvalue="page" event="click">
      <h2>BLOG 2</h2>
    </router-link-stub>
    <article>Blog content 2...</article>
  </div>
  <div class="single-blog">
    <router-link-stub to="/blog/undefined" tag="a" ariacurrentvalue="page" event="click">
      <h2>BLOG 3</h2>
    </router-link-stub>
    <article>Blog content 3...</article>
  </div>
</div>
`;

exports[`Showblogs list of blogs search by title into an empty blogs list returns list with 2 items 1`] = `
<div id="show-blogs">
  <h1>All Blog Articles</h1> <input type="text" placeholder="search blogs">
  <div class="single-blog">
    <router-link-stub to="/blog/undefined" tag="a" ariacurrentvalue="page" event="click">
      <h2>ANOTHER BLOG 1</h2>
    </router-link-stub>
    <article>Blog content 1...</article>
  </div>
  <div class="single-blog">
    <router-link-stub to="/blog/undefined" tag="a" ariacurrentvalue="page" event="click">
      <h2>ANOTHER BLOG 2</h2>
    </router-link-stub>
    <article>Blog content 2...</article>
  </div>
</div>
`;
